### DISCUSSION TITLE
Glued Contact Convergence with multiple Contacts

### DISCUSSION THREAD

**Discussion by makeclean** [idaholab/moose#19130] [Q&A General] [answered]
Hi All, y'all must feel like I'm bombarding you! I've got a problem with a contact convergence problem, I've tried making an MVP type problem, but it converges fine, so I must be missing something in my simplification. I have the following assembly; 

![Untitled presentation](https://user-images.githubusercontent.com/2439754/137681717-9ce333ac-d315-452d-beba-71bd49dbb484.png)

I've made the problem smaller, e.g. modelling the contact between the ball and the pin with glued contact, and that converges ok. I've also modelled the contact between the ball and the 'hat' and that converges fine (my initial guess was that it was a separation issue between the ball and the hat, that turned out to be wrong). I've also constrained the 'hat' to move in z only but that made no appreciable difference. So independently pin-ball OK, ball-hat OK, pin-ball-hat - NOPE. 

Any suggestions?

**Comment by GiudGiud** [comment]
@bwspenc @lindsayad @recuero

**Comment by recuero** [comment]
I cannot distinguish the geometry very well from the image. But, from the description, it seems that the determining factor there is the application of the glued contact constraints altogether. I'd make sure there is no part of the mesh that gets overconstrained (maybe reducing the size of some sidesets?).

**Reply by makeclean** [reply]
Weirdly, there are some points where the penetration is -1e6, i.e. -ve 1 million metres which doesn't make much sense to me frankly.

![Screenshot from 2021-10-19 13-53-08](https://user-images.githubusercontent.com/2439754/137913339-a0f6d83a-ef5a-4336-9783-1e8cfa3ae161.png)

Also, when attempting to converge there is a bail out, looks like somehow one of the elements is either displacing or the nodes are shooting off to infinite +z
```

Setting Up
  Initializing
    Finished Initializing Equation Systems                                               [  4.84 s] [  807 MB]
    Finished Initializing Displaced Equation System                                      [  0.60 s] [  386 MB]
  Finished Initializing                                                                  [  6.03 s] [ 1196 MB]
Finished Setting Up                                                                      [  7.51 s] [ 1375 MB]
Framework Information:
MOOSE Version:           git commit b5a350860f on 2021-09-28
LibMesh Version:         aebb5a5c0e1f6d8cf523a720e19f70a6d17c0236
PETSc Version:           3.15.1
Current Time:            Tue Oct 19 19:23:06 2021
Executable Timestamp:    Tue Sep 28 20:19:26 2021

Parallelism:
  Num Processors:          1
  Num Threads:             8

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   35424
  Elems:                   183877
  Num Subdomains:          3

Nonlinear System:
  Num DOFs:                106272
  Num Local DOFs:          106272
  Variables:               { "disp_x" "disp_y" "disp_z" } 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Auxiliary System:
  Num DOFs:                6025760
  Num Local DOFs:          6025760
  Variables:               { "stress_xx_nodal" "strain_xx_nodal" "stress_yy_nodal" "strain_yy_nodal" 
                             "stress_zz_nodal" "strain_zz_nodal" "vonmises_nodal" } { "strain_xx" "strain_yy" 
                             "strain_zz" "vonmises_stress" } { "penetration" "contact_pressure" "nodal_area_contact_ball_socket" 
                             "nodal_area_contact_pin_ball" } 
  Finite Element Types:    "MONOMIAL" "MONOMIAL" "LAGRANGE" 
  Approximation Orders:    "FIRST" "CONSTANT" "FIRST" 

Execution Information:
  Executioner:             Transient
  TimeStepper:             ConstantDT
  Solver Mode:             NEWTON
  MOOSE Preconditioner:    SMP


    Setting Up Materials
      Finished Computing Initial Material Values                                         [  3.40 s] [  997 MB]
    Finished Setting Up Materials                                                        [  3.41 s] [  998 MB]
    Reinitializing Because of Geometric Search Objects..                                 [ 25.42 s] [  141 MB]
  Finished Performing Initial Setup                                                      [ 32.53 s] [ 1157 MB]

Time Step 0, time = 0

Time Step 1, time = 5, dt = 5

    Rebuilding Geometric Search Patches
      Reinitializing Because of Geometric Search Objects...                              [ 26.22 s] [  -25 MB]
    Finished Rebuilding Geometric Search Patches                                         [ 26.34 s] [  -24 MB]

Performing automatic scaling calculation

 0 Nonlinear |R| = 3.216429e-07
    Computing Jacobian                                                                   [  7.49 s] [  198 MB]
      0 Linear |R| = 3.216429e-07
      1 Linear |R| = 3.190770e-07
      2 Linear |R| = 3.173445e-07
      3 Linear |R| = 1.025635e-13
WARNING: At least one element took more than 10 iterations to converge in inverse_map()...
Rerun in devel/dbg mode for more details.
ERROR: Newton scheme FAILED to converge in 21 iterations in element 122391 for physical point = (x,y,z)=(-0.0166628,  1.63084, 4.32143e+08)
  Elem Information
   id()=122391, unique_id()=198603, processor_id()=0
   type()=TET4
   dim()=3
   n_nodes()=4
    0  Node id()=24809, processor_id()=0, Point=(x,y,z)=(-0.0165868,   1.6312, 4.32143e+08)
    DoFs=(0/0/74427) (0/1/74428) (0/2/74429) (1/11/5983300) (1/12/5983301) (1/13/5983302) (1/14/5983303) 
    1  Node id()=24811, processor_id()=0, Point=(x,y,z)=(-0.0167855,  1.62849, 4.32143e+08)
    DoFs=(0/0/74433) (0/1/74434) (0/2/74435) (1/11/5983308) (1/12/5983309) (1/13/5983310) (1/14/5983311) 
    2  Node id()=23894, processor_id()=0, Point=(x,y,z)=(-0.0192472,  1.63069, 4.32143e+08)
    DoFs=(0/0/71682) (0/1/71683) (0/2/71684) (1/11/5979640) (1/12/5979641) (1/13/5979642) (1/14/5979643) 
    3  Node id()=23893, processor_id()=0, Point=(x,y,z)=(-0.0197317,  1.62942, 4.32143e+08)
    DoFs=(0/0/71679) (0/1/71680) (0/2/71681) (1/11/5979636) (1/12/5979637) (1/13/5979638) (1/14/5979639) 
   n_sides()=4
    neighbor(0)=nullptr
    neighbor(1)=126093
    neighbor(2)=137441
    neighbor(3)=128532
   hmin()=0.00303343, hmax()=0.00427406
   volume()=4.45414e-09
   active()=1, ancestor()=0, subactive()=0, has_children()=0
   parent()=nullptr
   level()=0, p_level()=0
   refinement_flag()=DO_NOTHING
   p_refinement_flag()=DO_NOTHING
   DoFs=(1/0/3426948) (1/0/3426949) (1/0/3426950) (1/0/3426951) (1/1/3426952) (1/1/3426953) (1/1/3426954) (1/1/3426955) (1/2/3426956) (1/2/3426957) (1/2/3426958) (1/2/3426959) (1/3/3426960) (1/3/3426961) (1/3/3426962) (1/3/3426963) (1/4/3426964) (1/4/3426965) (1/4/3426966) (1/4/3426967) (1/5/3426968) (1/5/3426969) (1/5/3426970) (1/5/3426971) (1/6/3426972) (1/6/3426973) (1/6/3426974) (1/6/3426975) (1/7/5638120) (1/8/5638121) (1/9/5638122) (1/10/5638123) 
```

**Reply by makeclean** [reply]
Oh I should also note, i turned on automatic scaling, which was off previously. Could the symmetry plane we can see make a difference?

**Reply by recuero** [reply]
Could be. I'd try doing glued contact with the penalty formulation if you haven't.

**Reply by lindsayad** [reply]
> Oh I should also note, i turned on automatic scaling, which was off previously. Could the symmetry plane we can see make a difference?

So did the large penetration only show up after turning on automatic scaling?

**Reply by makeclean** [reply]
With penalty, same deal as above. Without symmetry, I get -ve Jacobian error
```
[adavis@j0818 run]$ ~/opt/moose/modules/combined/combined-opt -i p5coil_ball_mount-sym.i --n-threads=8

Setting Up
  Finished Setting Mesh                                                                  [  0.78 s] [  136 MB]
  Initializing
    Initializing Equation Systems                                                        [  7.43 s] [ 1238 MB]
    Finished Initializing Displaced Equation System                                      [  0.92 s] [  570 MB]
  Finished Initializing                                                                  [  9.38 s] [ 1816 MB]
Finished Setting Up                                                                      [ 11.75 s] [ 2050 MB]
Framework Information:
MOOSE Version:           git commit b5a350860f on 2021-09-28
LibMesh Version:         aebb5a5c0e1f6d8cf523a720e19f70a6d17c0236
PETSc Version:           3.15.1
Current Time:            Tue Oct 19 20:51:23 2021
Executable Timestamp:    Tue Sep 28 20:19:26 2021

Parallelism:
  Num Processors:          1
  Num Threads:             8

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   50807
  Elems:                   272155
  Num Subdomains:          3

Nonlinear System:
  Num DOFs:                152421
  Num Local DOFs:          152421
  Variables:               { "disp_x" "disp_y" "disp_z" } 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Auxiliary System:
  Num DOFs:                8912188
  Num Local DOFs:          8912188
  Variables:               { "stress_xx_nodal" "strain_xx_nodal" "stress_yy_nodal" "strain_yy_nodal" 
                             "stress_zz_nodal" "strain_zz_nodal" "vonmises_nodal" } { "strain_xx" "strain_yy" 
                             "strain_zz" "vonmises_stress" } { "penetration" "contact_pressure" "nodal_area_contact_ball_socket" 
                             "nodal_area_contact_pin_ball" } 
  Finite Element Types:    "MONOMIAL" "MONOMIAL" "LAGRANGE" 
  Approximation Orders:    "FIRST" "CONSTANT" "FIRST" 

Execution Information:
  Executioner:             Transient
  TimeStepper:             ConstantDT
  Solver Mode:             NEWTON
  MOOSE Preconditioner:    SMP


    Setting Up Materials
      Finished Computing Initial Material Values                                         [  4.86 s] [ 1464 MB]
    Finished Setting Up Materials                                                        [  4.86 s] [ 1464 MB]
    Reinitializing Because of Geometric Search Objects.....                              [ 42.59 s] [  172 MB]
  Finished Performing Initial Setup                                                      [ 51.93 s] [ 1661 MB]

Time Step 0, time = 0

Time Step 1, time = 5, dt = 5

    Rebuilding Geometric Search Patches
      Reinitializing Because of Geometric Search Objects......                           [ 44.09 s] [    4 MB]
    Finished Rebuilding Geometric Search Patches                                         [ 44.30 s] [    4 MB]

Performing automatic scaling calculation

 0 Nonlinear |R| = 9.152297e-07
    Computing Jacobian.                                                                  [ 12.57 s] [  584 MB]
      0 Linear |R| = 9.152297e-07
      1 Linear |R| = 8.865040e-07
      2 Linear |R| = 8.865040e-07
      3 Linear |R| = 1.385801e-09
  Elem Information
   id()=244540, unique_id()=313416, processor_id()=0
   type()=TET4
   dim()=3
   n_nodes()=4
    0  Node id()=44934, processor_id()=0, Point=(x,y,z)=(-0.0438262,  1.75396, -3.93721e+09)
    DoFs=(0/0/134802) (0/1/134803) (0/2/134804) (1/11/8888696) (1/12/8888697) (1/13/8888698) (1/14/8888699) 
    1  Node id()=44935, processor_id()=0, Point=(x,y,z)=(-0.105596,  1.72867, -3.93721e+09)
    DoFs=(0/0/134805) (0/1/134806) (0/2/134807) (1/11/8888700) (1/12/8888701) (1/13/8888702) (1/14/8888703) 
    2  Node id()=44925, processor_id()=0, Point=(x,y,z)=(-0.0108594,  1.67695, -3.93721e+09)
    DoFs=(0/0/134775) (0/1/134776) (0/2/134777) (1/11/8888660) (1/12/8888661) (1/13/8888662) (1/14/8888663) 
    3  Node id()=44929, processor_id()=0, Point=(x,y,z)=(-0.00339524,  1.68348, -3.93721e+09)
    DoFs=(0/0/134787) (0/1/134788) (0/2/134789) (1/11/8888676) (1/12/8888677) (1/13/8888678) (1/14/8888679) 
   n_sides()=4
    neighbor(0)=246722
    neighbor(1)=246886
    neighbor(2)=245556
    neighbor(3)=247795
   hmin()=0.0226688, hmax()=0.135425
   volume()=-2.01603e-05
   active()=1, ancestor()=0, subactive()=0, has_children()=0
   parent()=nullptr
   level()=0, p_level()=0
   refinement_flag()=DO_NOTHING
   p_refinement_flag()=DO_NOTHING
   DoFs=(1/0/6847120) (1/0/6847121) (1/0/6847122) (1/0/6847123) (1/1/6847124) (1/1/6847125) (1/1/6847126) (1/1/6847127) (1/2/6847128) (1/2/6847129) (1/2/6847130) (1/2/6847131) (1/3/6847132) (1/3/6847133) (1/3/6847134) (1/3/6847135) (1/4/6847136) (1/4/6847137) (1/4/6847138) (1/4/6847139) (1/5/6847140) (1/5/6847141) (1/5/6847142) (1/5/6847143) (1/6/6847144) (1/6/6847145) (1/6/6847146) (1/6/6847147) (1/7/8598500) (1/8/8598501) (1/9/8598502) (1/10/8598503) 
ERROR: negative Jacobian -0.000120962 at point (x,y,z)=(-0.0409192,  1.71076, -3.93721e+09) in element 244540
```
Somehow the nodes are shooting off to -ve infinity in this case??

**Reply by makeclean** [reply]
Here are the regions (circled) of large -ve penetration, upper threshold is -0.01 
![Untitled presentation(1)](https://user-images.githubusercontent.com/2439754/137984084-2ab74c06-befd-48a3-a179-be91b7dd0b16.png)

**Reply by makeclean** [reply]
@lindsayad the -ve penetration is consistent between them, however in the case of automatic scaling=true I get a -ve jacobian error, in the case of automatic_scaling=false its the `newton scheme failed to converge`

**Reply by makeclean** [reply]
What I absolutely do not understand is why I would see large -ve penetration in that particular volume when it does not show up in the case when I model the ball and pin, in isolation without the 'hat'

**Reply by makeclean** [reply]
happy to share files here, if that would be useful

**Reply by bwspenc** [reply]
A common issue for secondary nodes that are right on the edge of primary faces (which appears to be where you are having trouble) is that if they are slightly off the face, contact isn't picked up. You can remedy that by setting the `tangential_tolerance` parameter, which controls a distance (in model coordinates) by which the faces are virtually extended off beyond the edge. I'd try setting that to something like 1/10 the length of your elements in that region.

I'm puzzled about why this makes the solver fall apart, because the worst case should really just be that contact isn't enforced on those nodes, but it's worth giving it a try.

**Reply by makeclean** [reply]
Ive tried numbers from 1e-6 to 1e-3 (mesh size) but no avail, Im now trying a 2d version to see if it has the same problem

**Reply by makeclean** [reply]
Scratch that I dont think 2d will show the right behaviour

**Reply by bwspenc** [reply]
You're using glued contact for both of the interaction pairs (pin-ball and ball-hat), right? Also, have you tried various combinations of primary/secondary for both of those interaction pairs?

**Reply by makeclean** [reply]
Yes glued for both, I haven't tried different combinations since individually (ball-pin) and (ball-hat) the contact pairs as defined in their independent runs converged fine. I'll give it a go though.

**Reply by makeclean** [reply]
Well.... switching the order of the contact surfaces for ball/pin made the simulation run to convergence, however...
![Screenshot from 2021-10-19 22-28-15](https://user-images.githubusercontent.com/2439754/137993371-fca83478-4cd8-47ab-82e1-38ae66dd2164.png)
I'm thinking theres a bit of over-penetration there...

**Reply by bwspenc** [reply]
Just a bit!

Now I'm wondering whether we have some sort of fundamental issue related to handling of chains of blocks that are constrained to each other. You mentioned at the beginning that you tried to set up some sort of minimal problem, which actually worked. Was it just something like 3 blocks next to each other with contact constraints between them?

I can't promise that I'll find the time to do much with this, but it would be helpful if you could share your model.

**Comment by makeclean** [comment, marked as answer]
I can send you this problem, along with my test problem. My original hypothesis, was the curvature of these surfaces came into play in some way so the test problem has a sphere, with conical sections of spherical shells above it. In that case three layers with contact between them was fine. They were exact contact where there was no gaps between the shells, tomorrow I can try a version with gaps.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by makeclean** [reply]
[problem-contact.tar.gz](https://github.com/idaholab/moose/files/7391969/problem-contact.tar.gz) - here is the explicit problem above

**Reply by makeclean** [reply]
[contact-mvp.tar.gz](https://github.com/idaholab/moose/files/7392026/contact-mvp.tar.gz) and the contct mvp - I have created and confirmed that there is a double contact problem which works fine, and this triple contact fails.

**Reply by makeclean** [reply]
Oh, I think I may've done something quite dim. I use a pressure boundary, which seems to work absolutely fine when there are two contacts, but not when there are three. However, if I use FunctionDirichletBC for displacement, it seems to be fine. The particular comparison we're doing with ANSYS uses force/pressure based BCs. So I've now figured out from your advice in #17780, what the right thing to do is, and now its working.  

![Screenshot from 2021-10-21 22-10-54](https://user-images.githubusercontent.com/2439754/138357268-cbeb8aaa-3b0f-45fa-8072-e5e2950b6508.png)

Thanks for your help, for posterity I'll upload a complete input & mesh which runs as I expect.

**Reply by recuero** [reply]
Glad it's working!

**Reply by bwspenc** [reply]
I'm horrible at monitoring my notifications -- sorry for not responding earlier.  I'm glad you got this working!  Yes, you have to be really careful about making sure you don't have unrestrained rigid body modes for problems like this.
